feat: science module

ejcheng opened the pull request:
closes #944

ejcheng committed:
feat: science module

ejcheng committed:
test: Science module tests

ST-DDT commented on src/definitions/science.ts:
```diff
@@ -0,0 +1,33 @@
+ * The possible definitions related to science.
+ */
+export type ScienceDefinitions = LocaleEntry<{
+  /**
+   * Some science units (short).
+   */
+  shortUnit: string[];
```
Maybe unitAbbr, as we already use that term elsewhere.

Shinigami92 replied:
How about structuring it like `unit: { full, abbr }`? That way, the data can be coupled.
SO you could use 

```ts
const { abbr, full } = science.unit()
```

ST-DDT replied:
That doesn't match up with our other methods like date.month and weekday.
And I dont see a usecase for it either.

Shinigami92 replied:
The use-case would be that you just want one unit and show it in a table or whatever
So you have e.g. `Take 200 g (gram) of salt`

Shinigami92 replied:
I may want to design the httpStatus call in a same way
But you are right as currently we don't have that yet (maybe currency?)
I wont gave a veto if it gets merged as is, but it would be a bit better UX in my eyes :shrug:

pkuczynski replied:
I agree with @Shinigami92 it would be better. Gave the same feedback to `element` in the #944. Otherwise you could not get both short and long name for the unit. And if user wants only one, he could do: `faker.unit().short`

ST-DDT commented on src/definitions/science.ts:
```diff
@@ -0,0 +1,33 @@
+   * Some science units (long).
+   */
+  longUnit: string[];
+  /**
+   * Some element informtion.
+   */
+  element: Element[];
```
I think we should explicitly refer to chemical elements. As `element` is kind of generic.

pkuczynski replied:
In science? What other elements in science there are?

ST-DDT commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,65 @@
+   * faker.science.unit({ long: false }) // 'Pa'
+   * faker.science.unit({ long: true }) // 'joule'
+   */
+  unit(options?: { long?: boolean }): string {
```
```suggestion
  unit(options: { long?: boolean } = {}): string {
```

ST-DDT commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,65 @@
+   */
+  unit(options?: { long?: boolean }): string {
+    const { long = true } = options || {};
+    options = { long };
```
```suggestion
```

ST-DDT commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,65 @@
+   * faker.science.unit({ long: true }) // 'joule'
+   */
+  unit(options?: { long?: boolean }): string {
+    const { long = true } = options || {};
```
```suggestion
    const { long = true } = options;
```

ejcheng committed:
Apply suggestions from @ST-DDT's code review

Co-authored-by: ST-DDT <ST-DDT@gmx.de>

ejcheng committed:
chore: rename element to chemicalElement

ejcheng committed:
Merge branch 'main' into science

pkuczynski commented on src/definitions/science.ts:
```diff
@@ -13,9 +13,9 @@ export type ScienceDefinitions = LocaleEntry<{
+   * Some chemical element informtion.
    */
-  element: Element[];
+  chemicalElement: Element[];
```
Is it worth changing the name? Do we have other `element` function we would like to develop? If we make the change, we should also change `Element` type to `ChemicalElement`, but I doubt the value in this change tbh.

Shinigami92 replied:
IMO yes, we should use `chemicalElement` and `ChemicalElement`

pkuczynski replied:
I am wondering if we should not use just `element` as what other elements there might be in the science module? Its even called `periodic table of elements` isn't it?

import-brain replied:
> I am wondering if we should not use just `element` as what other elements there might be in the science module? Its even called `periodic table of elements` isn't it?

Yeah, I can't recall any other thing commonly referred to as an "element" that's widely-used in science.

pkuczynski replied:
Me neither, that's why I would keep it as discussed initially: `element`. @ST-DDT @Shinigami92 wdyt?

ST-DDT replied:
I agree with you, that's it's the most common one in science.
But it's not the only one: https://en.wikipedia.org/wiki/Electrical_element
Even wikipedia refers to it as https://en.wikipedia.org/wiki/Chemical_element

Similar to `statusCode`s in the internet referring mostly to http status codes, but there are also other status codes such as `grpc`'s one.

Should we discuss this in one of our next team meetings?

pkuczynski replied:
Wikipedia is a bit different than us because they do not have namespaces, so they sometimes need a bit longer article names. The elements are equally chemical and physical, so I don't think this is a nice prefix. Electrical elements would probably be included in a different module I can imagine as the more for designing devices then pure scientific data, isn't it. Even in Wikipedia it says _periodic table of the (chemical) elements_ https://en.wikipedia.org/wiki/Periodic_table, where _chemical_ is in brackets...

pkuczynski commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,64 @@
+   * @example
+   * faker.science.elementName() // 'Praseodymium'
+   */
+  elementName(): string {
```
Haven't you agreed in #944 to implement it as `faker.element().name`?

import-brain replied:
*facepalm* Whoops, I forgot

pkuczynski commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,64 @@
+   * @example
+   * faker.science.elementSymbol() // 'Rg'
+   */
+  elementSymbol(): string {
```
`faker.element().symbol`

pkuczynski commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,64 @@
+   * faker.science.unit({ long: false }) // 'Pa'
+   * faker.science.unit({ long: true }) // 'joule'
+   */
+  unit(options: { long?: boolean } = {}): string {
```
`faker.unit().long`, `faker.unit().short`

ejcheng committed:
chore: chemicalElement -> ChemicalElement

ejcheng committed:
feat: return dictionary from element()

ejcheng committed:
refactor: refactor unit, doc changes

ejcheng committed:
refactor: original objs instead of making new ones

ST-DDT commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,58 @@
+   * faker.science.element() // { name: 'Cerium', symbol: 'Ce' }
+   */
+  element(): Element {
+    const index = this.faker.datatype.number({
+      min: 0,
+      max: this.faker.definitions.science.chemicalElement.length - 1,
+    });
+    return this.faker.definitions.science.chemicalElement[index];
```
Why don't you use `helpers.arrayElement`?
```suggestion
    return this.faker.helpers.arrayElement(this.faker.definitions.science.chemicalElement);
```

Shinigami92 commented on src/locales/en/science/unit.ts:
```diff
@@ -0,0 +1,90 @@
+  },
+  {
+    long: 'ohm',
+    short: 'Ω',
```
Should we also name `short` just `symbol`?

import-brain replied:
If we rename `short` to `symbol`, should we rename `long` to `full` or `name`?

Shinigami92 replied:
yes (😉 => _not sure whether `full` or `name`_)

pkuczynski replied:
I vote for `name`

ST-DDT replied:
I also prefer `name`.

import-brain replied:
Okay, changing to `name`

Shinigami92 commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,58 @@
+import type { Faker } from '../..';
+
+export type Element = {
+  name: string;
+  symbol: string;
+};
```
```suggestion
export interface Element {
  name: string;
  symbol: string;
};
```

Shinigami92 commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,58 @@
+  symbol: string;
+};
+
+export type Unit = {
+  long: string;
+  short: string;
+};
```
```suggestion
export interface Unit {
  long: string;
  short: string;
};
```

Shinigami92 reviewed (changes requested):
Remember to re-export the interfaces ChemicalElement and Unit in `src/index.ts`

ejcheng committed:
Merge branch 'main' into science

ejcheng committed:
Apply code review suggestions from @ST-DDT and @Shinigami92

Co-authored-by: Shinigami <chrissi92@hotmail.de>
Co-authored-by: ST-DDT <ST-DDT@gmx.de>

ejcheng committed:
chore: fix prettier linting issues

Shinigami92 commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,54 @@
+   * faker.science.unit() // { long: 'mole', short: 'mol' }
+   */
+  unit(): Unit {
+    const index = this.faker.datatype.number({
+      min: 0,
+      max: this.faker.definitions.science.unit.length - 1,
+    });
+    return this.faker.definitions.science.unit[index];
```
pls use `arrayElement` for such things

Shinigami92 commented on test/science.spec.ts:
```diff
@@ -0,0 +1,123 @@
+    for (let i = 1; i <= NON_SEEDED_BASED_RUN; i++) {
+      describe(`elementName()`, () => {
+        it('should return a string', () => {
+          const name = faker.science.elementName();
```
tests need updates

Shinigami92 commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,54 @@
+export interface Element {
+  name: string;
+  symbol: string;
+};
```
run `pnpm run format`

Shinigami92 replied:
`pnpm run clean && pnpm install && pnpm run format`

Shinigami92 replied:
is something broken with prettier?

Shinigami92 replied:
it's already solved :thinking:

ejcheng committed:
refactor: Element -> ChemicalElement

rename Element to ChemicalElement in src/modules/science/index.ts to
keep type naming consistent throughout science module

ejcheng committed:
chore: re-export ChemicalElement + Unit interfaces

Co-authored-by: Shinigami <chrissi92@hotmail.de>

ejcheng committed:
refactor: use helpers.arrayElement for unit()

Co-authored-by: Shinigami <chrissi92@hotmail.de>

ejcheng committed:
test: updated tests for new architecture

ejcheng committed:
docs: change object doc to account for key names

ejcheng committed:
refactor: `long` -> `name` and `short` -> `symbol`

ejcheng committed:
Merge branch 'main' into science

ST-DDT commented on src/modules/science/index.ts:
```diff
@@ -0,0 +1,52 @@
+import type { Faker } from '../..';
+
+export interface ChemicalElement {
+  name: string;
+  symbol: string;
+}
+
+export interface Unit {
+  name: string;
+  symbol: string;
+}
```
Should we reuse the types in the definitions (or vice versa)?

ejcheng committed:
test: make seeded tests run green

ejcheng committed:
Merge branch 'main' into science

ejcheng committed:
Merge branch 'main' into science

ejcheng committed:
refactor: `element()` -> `chemicalElement()`

Shinigami92 commented on src/locales/en/science/chemicalElement.ts:
```diff
@@ -0,0 +1,474 @@
+export default [
+  {
+    symbol: 'H',
```
Should we add a third field `atomicNumber`?

import-brain replied:
Sure. Should the number be a string or an number type?

Shinigami92 replied:
Mh... I think a number would fit here, or are there special cases?

ejcheng committed:
refactor: reuse types

import-brain commented on src/locales/en/science/chemicalElement.ts:
```diff
@@ -0,0 +1,474 @@
+export default [
```
By "freeze", do you mean use `Object.freeze()`?

ST-DDT replied:
Yes, so that unit().name = 123 doesnt do anything unexpected.

ejcheng committed:
Merge branch 'main' into science

ejcheng committed:
chore: freeze objects

ejcheng committed:
feat: add atomic number

ejcheng committed:
Merge branch 'main' into science

ejcheng committed:
test: add atomicNumber to seeded tests

Shinigami92 committed:
Merge branch 'main' into science

Shinigami92 committed:
Merge branch 'main' into science
